timings/lib/src/timings.dart

```diff
@@ -0,0 +1,386 @@
+// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:timings/src/clock.dart';
+
+/// The timings of an operation, including its [startTime], [stopTime], and
+/// [duration].
+abstract class TimeSlice {
+  /// The total duration of this operation, equivalent to taking the difference
+  /// between [stopTime] and [startTime].
+  Duration get duration => stopTime?.difference(startTime);
+
+  DateTime get startTime;
+
+  DateTime get stopTime;
+
+  TimeSlice._();
+
+  factory TimeSlice(DateTime startTime, DateTime stopTime) =>
+      _TimeSliceImpl(startTime, stopTime);
+
+  @override
+  String toString() => '($startTime + $duration)';
+}
+
+class _TimeSliceImpl extends TimeSlice {
+  @override
+  final DateTime startTime;
+
+  @override
+  final DateTime stopTime;
+
+  _TimeSliceImpl(this.startTime, this.stopTime) : super._();
+}
+
+/// The timings of an async operation, consist of several sync [slices] and
```

Comment:
do we make any guarantees about how `startTime` and `endTime` relate to the times in `slices`? Should we?